compose: Show avatars for people in typeahead autocompletes.

skunkmb opened the pull request:
This PR adds avatars in the typeahead autocompletes next to users. Users without avatars (e.g. `@all` and `@everyone`) have an empty placeholder instead so that they line up with the other names.

The `typeahead-image` class is similar to the `emoji` class also used by typeahead, but with a few differences in size, positioning, and its `border-radius`.

Fixes #6635.

**Screenshot:**

<img width="258" alt="screen shot 2018-03-31 at 11 01 54 am" src="https://user-images.githubusercontent.com/17259768/38166117-4b49fcb8-34d3-11e8-8e30-fa82245b4ff0.png">

timabbott commented:
@skunkmb I think we need something to go beside `@all` in this design; maybe try using a megaphone icon for the purpose?  Also, the vertical positioning looks a bit off; might be a screenshot artifact; can you check it?

skunkmb commented:
@timabbott OK, I made a megaphone icon and added it next to `@all`, so now it looks like this:

<img width="258" alt="screen shot 2018-03-31 at 2 07 16 pm" src="https://user-images.githubusercontent.com/17259768/38167615-dfb28456-34ec-11e8-870d-5c53c0ebf045.png">

As for the pictures being centered, I checked and they are. There are `4px` above and `4px` below, and `23px` to the left and `23px` to the right.

Please let me know if there is more I should fix. Thanks!

timabbott commented:
For the megaphone icon, check if there's something already in font awesome we can use?  That has perf advantages over making a new one.

timabbott commented:
(I think you can also probably squash the commits).

timabbott commented on static/js/typeahead_helper.js:
```diff
@@ -79,14 +79,20 @@ var rendered = { persons: new Dict(), streams: new Dict(), user_groups: new Dict
         html = exports.render_typeahead_item({
             primary: person.full_name,
             secondary: person.email,
+            img_src: 'avatar/' + person.email,
```
you should use the user ID here; both URLs work, but the email-based version is deprecated.

skunkmb replied:
@timabbott OK, I am now using the UID version, based on what is in `people.js`'s `small_avatar_url`. Please let me know if that was what you were referring to. Thanks!

skunkmb commented:
@timabbott OK, I used a FontAwesome megaphone icon:

<img width="230" alt="screen shot 2018-04-02 at 1 04 11 pm" src="https://user-images.githubusercontent.com/17259768/38213532-67f6b88e-3676-11e8-973b-ae9a81bdda2f.png">

I also squashed the commits. Please let me know if there is more to change. Thanks!

timabbott commented:
Can you fix the linter issue?  Also, what's the story with `messages.swift`?

timabbott commented on static/js/typeahead_helper.js:
```diff
@@ -79,14 +79,28 @@ var rendered = { persons: new Dict(), streams: new Dict(), user_groups: new Dict
+            avatar_url = 'https://secure.gravatar.com/avatar/' + hash + '?d=identicon';
+        }
+
+        var small_avatar_url = people.format_small_avatar_url(avatar_url)
```
I think we want all the logic around `gravatar` and things to live in `people.js` -- can you move this to a function there?  Might require some refactoring to avoid code duplication.

skunkmb replied:
@timabbott OK, I moved it to a separate function `small_avatar_url_for_person` and did some refactoring so that `small_avatar_url` uses that function as well if necessary. Thanks!

skunkmb commented:
@timabbott OK, the linter issue should be fixed.

As for `messages.swift`, I added that by accident! I've been tinkering around with making a Zulip library in Swift similar to the `zulip-js` library in my free time, because Swift is a language that I enjoy using and have worked in before. That was one of the test files. I can show it to you when I'm done if you're interested. I've removed it from this PR, though. Sorry about any confusion.

Please let me know if there is more to change or fix. Thanks!

rishig commented:
what do we do for @mobile or other user groups?

Also this is unrelated, but probably `@everyone` and `@all` should have `(Notify stream)` instead of `(Notify everyone)` in the parenthesis.

timabbott commented:
Good point, I'll make that change to the text (also need to fix it missing i18n tags) while we're thinking about it.

For user groups, I think we probably just want the icon we use for user groups in settings (I think it's the multiple people icon).

timabbott commented:
Fixed the text thing in 721b4e83733b0e5f569737cb1469deb1526b9af8.  I think we can test-deploy this on chat.zulip.org once the user-groups thing is handled.

skunkmb commented:
@rishig @timabbott OK, user groups have an icon now:

<img width="291" alt="screen shot 2018-04-04 at 8 29 14 pm" src="https://user-images.githubusercontent.com/17259768/38346008-38164ec2-3847-11e8-8f14-5162d799bf49.png">

Please let me know if there is more to change. Thanks!

timabbott commented:
Looks great!  I'm test-deploying on chat.zulip.org.

I'm not happy with the `people.js` refactor commit, and split it into 5 commits (with a slightly different result) in what I'm test-deploying on chat.zulip.org.

timabbott commented:
@skunkmb I merged #8985; you can rebase this on top of that now :)

timabbott commented:
OK, given the feedback, I went ahead and merged this.  Thanks @skunkmb!  

I opened https://github.com/zulip/zulip/issues/9041 for a follow-up to this.
